doc/api/worker_threads.md

```diff
@@ -242,6 +271,26 @@ new Worker('process.env.SET_IN_WORKER = "foo"', { eval: true, env: SHARE_ENV })
   });
 ```
 
+## `worker.setEnvironmentData(key[, value])`
+<!--YAML
+added: REPLACEME
+-->
+
+> Stability: 1 - Experimental
+
+* `key` {any} Any arbitrary, cloneable JavaScript value that can be used as a
+  {Map} key.
+* `value` {any} Any arbitrary, cloneable JavaScript value that will be cloned
+  and passed automatically to all new `Worker` instances. If `value` is passed
+  as `undefined`, any previously set value for the `key` will be deleted.
+
+The `worker.setEnvironmentData()` API sets the content of the
+`worker.environmentData` in all new `Worker` instances spawned from the current
+context.
+
+Calling `worker.setEnvironmentData()` will have no impact on the value of
+`worker.environmentData` on existing threads.
```

Comment:
does this mean the feature does not apply hierarchically to workers? i.e., a worker inheriting a set of data from its parent, and it augments it with another set that it wants to propagate to all the workers it spawns?

Reply:
Correct. It would need to be set within each context explicitly. It is not automatically inherited. 

Should it be? (Legitimate question)

Reply:
I think from general programming semantics around parent-child association and from user's angle, that may be an expectation. But I would like to hear from @addaleax / @coreyfarrell on what they think.

Reply:
Referring back to #30992 what will the expectation be for code injected to workers?  I lean towards thinking worker code injection and `setEnvironmentData` should both inherit or neither inherit.  This pushes me slightly in the direction of preferring inheritance as it could reduce the chances of an unstable `--require` order among threads.

Reply:
hmm... ok. I'm generally not a fan of inheriting context like that but I can see the argument. @addaleax, any thoughts?

Reply:
Yeah, I also think that this should definitely be inherited, partly because of what @gireeshpunathil said – this is what I’d personally expect here – but also, because from a practical point of view, it’s easier to handle the case where the data is present but not needed rather than the case where the data is not needed but present.

Reply:
> rather than the case where the data is not needed but present.

I am wondering you might have meant: `rather than the case where the data is needed but not present.`?

Reply:
Works for me. I'll update it soon to propagate the data.

Reply:
Would it make sense to remove `worker.setEnvironmentData` and have stuff use `worker.environmentData.set` instead?

Reply:
Well, given that `worker.environmentData` can be replaced (e.g. `worker.environmentData = undefined`, I don't think that would work. The `worker.setEnvironmentData()` would silently make sure that the value is a `Map` in case it has been changed.

Reply:
Alright... so, a question on the inheritance....

Should calling `worker_threads.setEnvironmentData()` update the *current value* of `worker_threads.environmentData` for the current thread?

So, for instance,

```js
const {
  Worker,
  isMainThread,
  environmentData,
  setEnvironmentData,
} = require('worker_threads');

if (isMainThread) {
  console.log(environmentData); // null initially
  setEnvironmentData('foo', 'bar');
  console.log(environmentData.get('foo')); // ???
  new Worker(__filename);
} else {
  setEnvironmentData('abc', 123);
  console.log(environmentData.get('foo')); // bar
  console.log(environmentData.get('abc')); // ???
}
```

The answer to this one is important because it carries a poitential bit of complexity and a performance penalty due to additional copying that needs to happen if `worker_threads.setEnvironmentData()` does *not* impact the current value of `worker_threads.environmentData`

Also, when we say "inherit", we have to be clear what happens in the following case:

```js
const someObject = worker_threads.environmentData.get('foo');
someObject.a.b.c = 1;
```

Would new worker threads pick up the modified version of `someObject` or would it receive the original version?

Reply:
Ping @coreyfarrell @addaleax @gireeshpunathil .. :-)

Reply:
> Well, given that `worker.environmentData` can be replaced (e.g. `worker.environmentData = undefined`, I don't think that would work. The `worker.setEnvironmentData()` would silently make sure that the value is a `Map` in case it has been changed.

Sorry for my slow responses.  Do we actually want `worker.environmentData` to be writable or is it currently just writable by default?  Being able to count on it always being a Map object would be nicer than having to type check it before calling the `get` method.

Reply:
**according to me**, the state of the environment data at the time of the `inheritance` is what  needs to be inherited - essentially, the modified version of the object.

Reply:
Well, the key question is whether calling `setEnvironmentData()` changes the value for the *current* thread, not the children. Sorry if I didn't make that clear.

Reply:
> Well, the key question is whether calling `setEnvironmentData()` changes the value for the *current* thread, not the children. Sorry if I didn't make that clear.

I'm fine with it setting the value for the current thread, I suspect this will avoid some complexity in the implementation and documentation.